### PR TITLE
Remove deprecated grpc option from pkg/kvstore/etcd_debug.go

### PR DESCRIPTION
**Pull request by davchos**
Remove deprecated grpc option from pkg/kvstore/etcd_debug.go

  Related:#32274

  Signed-off-by: David CHOSROVA <dchosrova@gmail.com>

### CODE REVIEW

**Review by giorio94 — CHANGES_REQUESTED** [review]
Thanks @davchos for the PR.

Unfortunately, removing that option alone modifies the behavior of the troubleshooting tool. For context, the goal of this specific snippet is to determine whether it is possible to correctly connect to the configured etcd instance, and report an error otherwise. Currently, the check is performed in two steps, first to assert that it is actually possible to establish the connection, and then to verify that it is possible to retrieve a given key (as a basic authorization check). Removing the `WithBlock` option leads the first check to succeed even if the connection cannot be established (with the subsequent error being possibly confusing), as well as causes the `DialTimeout` to no longer be effective (hence falling back to the global timeout, which is 5 seconds by default). 

To preserve a similar behavior as the current one with minimal code changes, I'd suggest to:
* Check upon `cl.Get` failure whether `cl.ActiveConnection().GetState() == connectivity.TransientFailure`, and in that case output the `Failed to establish connection` error message, rather than `Failed to retrieve key from etcd`.
* Create a dedicated subcontext with a 1 second timeout for the `cl.Get` operation, so that it terminates early in case of failures;
* Drop the `DialTimeout` setting, as no longer effective;

**Comment by davchos** [comment]
> Thanks @davchos for the PR.
> 
> Unfortunately, removing that option alone modifies the behavior of the troubleshooting tool. For context, the goal of this specific snippet is to determine whether it is possible to correctly connect to the configured etcd instance, and report an error otherwise. Currently, the check is performed in two steps, first to assert that it is actually possible to establish the connection, and then to verify that it is possible to retrieve a given key (as a basic authorization check). Removing the `WithBlock` option leads the first check to succeed even if the connection cannot be established (with the subsequent error being possibly confusing), as well as causes the `DialTimeout` to no longer be effective (hence falling back to the global timeout, which is 5 seconds by default).
> 
> To preserve a similar behavior as the current one with minimal code changes, I'd suggest to:
> 
> * Check upon `cl.Get` failure whether `cl.ActiveConnection().GetState() == connectivity.TransientFailure`, and in that case output the `Failed to establish connection` error message, rather than `Failed to retrieve key from etcd`.
> * Create a dedicated subcontext with a 1 second timeout for the `cl.Get` operation, so that it terminates early in case of failures;
> * Drop the `DialTimeout` setting, as no longer effective;

Hi @giorio94 ,

Thanks for the feedback and explanation, I make the changes.

David

File: pkg/kvstore/etcd_debug.go [hunk 1]
```diff
@@ -81,21 +82,26 @@ func EtcdDbg(ctx context.Context, cfgfile string, dialer EtcdDbgDialer, w io.Wri
...
 	cfg.Context = ctx
 	cfg.Logger = zap.NewNop()
-	cfg.DialOptions = append(cfg.DialOptions, grpc.WithBlock(), grpc.WithContextDialer(dialer.DialContext))
-	cfg.DialTimeout = 1 * time.Second // The client hangs in case the connection fails, hence set a short timeout.
+	cfg.DialOptions = append(cfg.DialOptions, grpc.WithContextDialer(dialer.DialContext))
 
 	cl, err := client.New(*cfg)
 	if err != nil {
-		iiw.Println("❌ Failed to establish connection: %s", err)
+		iiw.Println("❌ failed to create gRPC client: %s", err)
```
Comment (giorio94) [inline pkg/kvstore/etcd_debug.go]:
```suggestion
		iiw.Println("❌ Failed to create etcd client: %s", err)
```

File: pkg/kvstore/etcd_debug.go [hunk 2]
```diff
@@ -81,21 +82,26 @@ func EtcdDbg(ctx context.Context, cfgfile string, dialer EtcdDbgDialer, w io.Wri
...
 		return
 	}
 	defer cl.Close()
 
 	// Try to retrieve the heartbeat key, as a basic authorization check.
 	// It doesn't really matter whether the heartbeat key exists or not.
```
Comment (giorio94) [inline pkg/kvstore/etcd_debug.go:95]: Could you please update the comment to mention the fact that `client.New` does not block on connection failure, and hence we need to check the connection state to determine the type of failure?

**Review by giorio94 — COMMENTED** [review]
Thanks, just a couple of final nits inline. 
Could you please also squash the two commits into one?

**Review by giorio94 — COMMENTED** [review]
Thanks, the patch looks good to me. Last thing before running the tests and approving, could you please generalize the commit message to describe the actual change, rather than the review modifications?

**Comment by davchos** [comment]
> Thanks, the patch looks good to me. Last thing before running the tests and approving, could you please generalize the commit message to describe the actual change, rather than the review modifications?

Thanks for walking me through that change, I appreciate.

**Review by giorio94 — APPROVED** [review]
Thanks!

(You'll need to rebase onto main to make CI happy)

**Pull request merged**